token-lending/program/src/processor.rs

```diff
@@ -1176,6 +1184,363 @@ fn process_accrue_interest(program_id: &Pubkey, accounts: &[AccountInfo]) -> Pro
+        if &dex_market_pubkey != dex_market_info.key {
+            msg!("Invalid dex market account input");
+            return Err(LendingError::InvalidAccountInput.into());
+        }
+    }
+    if let COption::Some(dex_market_pubkey) = withdraw_reserve.dex_market {
+        if &dex_market_pubkey != dex_market_info.key {
+            msg!("Invalid dex market account input");
+            return Err(LendingError::InvalidAccountInput.into());
+        }
+    }
+
+    let mut obligation = Obligation::unpack(&obligation_info.data.borrow())?;
+    if obligation_info.owner != program_id {
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+
+    // @TODO: should this be comparing with deposit_or_withdraw_reserve_collateral_supply_info.key?
+    if &obligation.collateral_reserve != withdraw_reserve_info.key {
+        msg!("Invalid deposit or withdraw reserve account");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    let obligation_token_mint = unpack_mint(&obligation_token_mint_info.data.borrow())?;
+    if &obligation.token_mint != obligation_token_mint_info.key {
+        msg!("Obligation token mint input doesn't match existing obligation token mint");
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    let obligation_token_input = Token::unpack(&obligation_token_input_info.data.borrow())?;
+    if obligation_token_input_info.owner != token_program_id.key {
+        return Err(LendingError::InvalidTokenOwner.into());
+    }
+    if &obligation_token_input.mint != obligation_token_mint_info.key {
+        return Err(LendingError::InvalidTokenMint.into());
+    }
+
+    // accrue interest and update rates
+    assert_last_update_slot(&borrow_reserve, clock.slot)?;
+    assert_last_update_slot(&withdraw_reserve, clock.slot)?;
+
+    obligation.accrue_interest(borrow_reserve.cumulative_borrow_rate_wads)?;
+
+    let obligation_collateral_amount = obligation.deposited_collateral_tokens;
+    if obligation_collateral_amount == 0 {
+        return Err(LendingError::ObligationEmpty.into());
+    }
+
+    // @TODO: can these next checks be combined? e.g. obligation_collateral_amount <= collateral_amount
+    if obligation_collateral_amount < collateral_amount {
+        // @TODO: does this need a different error type?
+        return Err(LendingError::ObligationEmpty.into());
+    }
+
+    if obligation_collateral_amount == collateral_amount {
+        // @TODO: do obligations ever allow withdrawing all collateral?
+        return Err(LendingError::ObligationEmpty.into());
+    }
+
+    // Check obligation health
+    let trade_simulator = TradeSimulator::new(
+        dex_market_info,
+        dex_market_orders_info,
+        memory,
+        &lending_market.quote_token_mint,
+        // @TODO: are these mint accounts correct?
+        &borrow_reserve.liquidity.mint_pubkey,
+        &withdraw_reserve.liquidity.mint_pubkey,
+    )?;
```

Comment:
Hmm this trade simulator feels a little awkward for withdraws. Instead of trading currencies, we really only need to judge the health of the obligation by getting the best price like we do for liquidation. Probably better as a separate change, but any ideas on how we could better split the trade simulator from just fetching the best price?